### PR TITLE
Pure generics

### PR DESCRIPTION
**Pull request by zgrannan**
Started encoding for generics. Works in pure code, impure casts and traits are not yet supported.

### CODE REVIEW

File: prusti-encoder/src/encoders/type/lifted_generic.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+use prusti_rustc_interface::middle::ty;
+use task_encoder::{OutputRefAny, TaskEncoder};
+use vir::with_vcx;
+
+use crate::encoders::GenericEnc;
+
+/// Lifting of a Rust type parameter to a Viper value
+/// Internally, this is represented as a LocalDecl, because
+/// unsubstituted generic parameters will always corresponds
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/lifted_generic.rs]:
```suggestion
/// unsubstituted generic parameters will always correspond
```

File: prusti-encoder/src/encoders/type/lifted.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    type TaskKey<'tcx> = Self::TaskDescription<'tcx>;
+
+    type OutputRef<'vir> = LiftedTy<'vir>;
+
+    type OutputFullLocal<'vir> = ();
+
+    type OutputFullDependency<'vir> = ();
+
+    type EnqueueingError = ();
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/lifted.rs]: Let's omit these when it's the default for the associated type anyway.

File: prusti-encoder/src/encoders/pure_generic_cast.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
+use crate::encoders::rust_ty_generic_cast::RustTyGenericCastEnc;
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/pure_generic_cast.rs]: Please add a high-level description of the encoder.

File: prusti-encoder/src/encoders/type/domain.rs [hunk 4]
```diff
@@ -63,21 +63,30 @@ pub enum DiscrBounds<'vir> {
 
 #[derive(Clone, Copy, Debug)]
 pub enum DomainEncSpecifics<'vir> {
-    Param,
     Primitive(DomainDataPrim<'vir>),
     // structs, tuples
     StructLike(DomainDataStruct<'vir>),
-    EnumLike(Option<DomainDataEnum<'vir>>)
+    EnumLike(Option<DomainDataEnum<'vir>>),
+    Param,
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/domain.rs]: (I understand there are many reformatting changes in the PR, but is this also something from auto-format?)

File: prusti-encoder/src/encoders/type/domain.rs [hunk 5]
```diff
@@ -63,21 +63,30 @@ pub enum DiscrBounds<'vir> {
...
 }
 
 #[derive(Clone, Debug)]
 pub struct DomainEncOutputRef<'vir> {
     pub base_name: String,
-    pub domain: vir::DomainIdentUnknownArity<'vir>,
+    pub domain: vir::DomainIdent<'vir, KnownArityAny<'vir, DomainParamData<'vir>, 0>>,
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/domain.rs]: For later: should we have a type alias `NullaryArityAny` as well? The `0` threw me off for a second.

File: prusti-encoder/src/encoders/type/generic_cast.rs [hunk 6]
```diff
@@ -0,0 +1,191 @@
+use prusti_rustc_interface::middle::ty;
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/generic_cast.rs]: Please add a high-level description for the encoder.

File: prusti-encoder/src/encoders/type/generic_cast.rs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        /// Casts a generic type to a concrete type
+        make_concrete: vir::FunctionIdent<'vir, UnaryArity<'vir>>,
+    },
+    AlreadyGeneric,
+}
+
+impl<'vir> GenericCastOutputRef<'vir> {
+    /// Converts the snapshot `snap` to a generic "Param" snapshot, if it's not
+    /// encoded as one already.
+    pub fn cast_to_generic_if_necessary<'tcx, Curr, Next>(
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/generic_cast.rs]: Unsure about the `_if_necessary` and `_if_possible` suffixes in the method names. I understand what they do, but `_if_possible` at a callsite sounds very similar to a `try_` method in Rust, so I would probably expect an `Option` return or similar. I think this encoder is here to hide the fact that casting may or may not happen.

File: prusti-encoder/src/encoders/type/generic_cast.rs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        (
+            Self::EncodingError,
+            Option<Self::OutputFullDependency<'vir>>,
+        ),
+    > {
+        if ty.is_generic() {
+            deps.emit_output_ref::<Self>(*ty, GenericCastOutputRef::AlreadyGeneric);
+            return Ok((&[], ()));
+        }
+        assert!(!ty.is_generic());
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/generic_cast.rs]: Remove? This is either a comment or *maybe* a `debug_assert!`, but the `ty.is_generic()` check happens literally 4 lines up.

File: prusti-encoder/src/encoders/type/generic_cast.rs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            let base_name = domain_ref.base_name;
+            let num_ty_args = domain_ref.type_function.arity().len();
+
+            let mk_type_spec = |param| {
+                let lifted_param_snap_ty = generic_ref.param_type_function.apply(vcx, [param]);
+                if num_ty_args > 0 {
+                    let typs = (0..num_ty_args).map(|i| {
+                        vcx.mk_local_decl(vcx.alloc(format!("t{i}")), generic_ref.type_snapshot)
+                    });
+                    vcx.mk_exists_expr(
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/generic_cast.rs]: As we discussed, this should not be an existential quantifier.

File: prusti-encoder/src/encoders/type/lifted_generic.rs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+impl TaskEncoder for LiftedGenericEnc {
+    task_encoder::encoder_cache!(LiftedGenericEnc);
+
+    type TaskDescription<'tcx> = &'tcx ty::ParamTy;
+
+    type TaskKey<'tcx> = Self::TaskDescription<'tcx>;
+
+    type OutputRef<'vir> = LiftedGeneric<'vir>;
+
+    type OutputFullLocal<'vir> = ();
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/lifted_generic.rs]: Same comment about the defaults.

File: prusti-encoder/src/encoders/type/lifted_generic.rs [hunk 11]
```diff
@@ -0,0 +1,70 @@
+use prusti_rustc_interface::middle::ty;
+use task_encoder::{OutputRefAny, TaskEncoder};
+use vir::with_vcx;
+
+use crate::encoders::GenericEnc;
+
+/// Lifting of a Rust type parameter to a Viper value
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/lifted_generic.rs]: Please expand this description a bit. The "lifting" is an expression/local of type `Type`, representing one of the type parameters of a function.

File: prusti-encoder/src/util.rs [hunk 12]
```diff
@@ -0,0 +1,124 @@
+use prusti_rustc_interface::{
```
Comment (Aurel300) [inline prusti-encoder/src/util.rs]: Please rename this file. It is very tempting to start putting random miscellaneous functions into a `util.rs` file when one exists. For now everything here seems to be related to types at least. Could it go into the `type` module? Could this be called `most_generic_ty.rs`?

File: vir/src/data.rs [hunk 13]
```diff
@@ -10,6 +10,7 @@ pub struct LocalData<'vir> {
     pub ty: Type<'vir>,
 }
 
+#[derive(Eq, PartialEq)]
```
Comment (Aurel300) [inline vir/src/data.rs:16]: (Why were these needed?)

**Review by Aurel300 — CHANGES_REQUESTED** [review]
Mostly looks good, missing some documentation and the existential in the conversion functions should still be changed. In future PRs, please try to avoid the reformatting changes as much as possible. We can do an fmt PR at some point (or even do these file by file), but for review these changes add a lot of noise.

File: prusti-encoder/src/encoders/type/domain.rs (hunk 4 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/domain.rs]: The `Param` here is to indicate that a generic type was passed to `DomainEnc` in the `DomainEncSpecifics`. Unrelated to formatting changes.

File: prusti-encoder/src/encoders/type/generic_cast.rs (hunk 7 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/generic_cast.rs]: >  I think this encoder is here to hide the fact that casting may or may not happen.

Indeed. I can't think of much better names though (`maybe_cast_generic` I think would be unclear under what conditions the cast would occur). Perhaps simply `make_concrete` and `make_generic`, although slightly misleading, might work (`make_concrete` wouldn't always return a concrete type).

File: prusti-encoder/src/encoders/type/lifted_generic.rs (hunk 10 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/lifted_generic.rs]: `OutputFullLocal` doesn't have a default (perhaps it should?)

File: vir/src/data.rs (hunk 13 above)
Comment (zgrannan) [reply vir/src/data.rs:16]: It's necessary for [this function](https://github.com/zgrannan/prusti-dev/blob/9a110867ba73143343b6dd60bc0a864e8782e649/prusti-encoder/src/encoders/type/lifted.rs#L25-L40), which extracts the unique type parameters that should be used to instantiate the type, removing duplicate instances of the same parameter. For example, from type `Tuple3<T, U, Result<T, W>>` it would return `[T, U, W]`.

File: prusti-encoder/src/encoders/type/lifted_generic.rs (hunk 11 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/lifted_generic.rs]: updated!

File: prusti-encoder/src/encoders/type/generic_cast.rs (hunk 6 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/generic_cast.rs]: added

File: prusti-encoder/src/encoders/type/domain.rs (hunk 5 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/domain.rs]: Added it

File: prusti-encoder/src/encoders/type/generic_cast.rs (hunk 8 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/generic_cast.rs]: removed

File: prusti-encoder/src/encoders/type/generic_cast.rs (hunk 9 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/generic_cast.rs]: updated

**Comment by zgrannan** [comment]
@Aurel300 addressed your changes, also identified some missed places, mostly resulting in the additional encoder here: https://github.com/zgrannan/prusti-dev/blob/52f6a7e40153c6fe21f12cfcc5e3a0fb80af5006/prusti-encoder/src/encoders/aggregate_snap_args_cast.rs

File: prusti-encoder/src/encoders/type/domain.rs (hunk 4 above)
Comment (Aurel300) [reply prusti-encoder/src/encoders/type/domain.rs]: Oh the reason I'm asking is because the diff shows that the `Param` variant was moved to be the last one declared instead of the first one, but it already existed.

Comment (zgrannan) [reply prusti-encoder/src/encoders/type/domain.rs]: Ah now I remember, I think what happened was I deleted it because I thought it was no longer necessary, and then I re-added it sometime later (at the bottom)

File: prusti-encoder/src/encoders/pure_generic_cast.rs (hunk 3 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/pure_generic_cast.rs]: done

**Comment by Aurel300** [comment]
Thanks! How did you test the implementation? I tried locally with this:

```rust
use prusti_contracts::*;

#[pure] fn id<T>(a: T) -> T { a }
#[pure] fn test1() -> i32 { id(42) }
#[pure] fn test2() -> bool { id(true) }

#[ensures(test1() == 42)]
#[ensures(test2())]
fn main() {}
```

(Which crashes, I added a `mir::TerminatorKind::Drop { .. } => { stmt_update }` case in `mir_pure.rs` to "handle" the `Drop` resulting from the `T` type.)

This should eventually generate a single definition for `id` with type parameter arguments, rather than the two `CALLER` variants. I guess this PR is handling type casts only, but not yet pure function calls?

**Comment by zgrannan** [comment]
@Aurel300 I tested it with `local-testing/test.rs` and some ad-hoc tests that I lost; so it wasn't very systematic. I'll add some tests and put them in local-testing.

Is it okay to include the drop handling in that way for this PR? I think it should be sound, since there is no way to have pre- or postconditions on drop without extern specs. **I added drop handling in https://github.com/Aurel300/prusti-dev/pull/39/commits/e9843af367d2927e17e58d3b7b484c530096bf04**

**Comment by zgrannan** [comment]
> I guess this PR is handling type casts only, but not yet pure function calls?

Yes ~~, in the sense that it doesn't handle the impure version of pure calls (for which CALLER variants would be generated)~~ (indeed, it doesn't handle the pure calls).

**Comment by zgrannan** [comment]
@Aurel300 refactored a bit, added some tests, and support for pure function calls

File: prusti-encoder/src/lib.rs [hunk 14]
```diff
@@ -79,14 +84,38 @@ pub fn test_entrypoint<'tcx>(
...
+            Vec<DomainFunction<'a>>,
+            Vec<DomainAxiom<'a>>
+        );
+
+        impl <'a> Debug for DomainWithExtras<'a> {
+            fn fmt<'b>(&self, f: &mut Formatter<'_>) -> fmt::Result {
+                fmt_domain_with_extras(f, self.0, &self.1, &self.2)
+            }
+        }
+        viper_code.push_str(&format!("{:?}\n", DomainWithExtras(output, type_functions, type_axioms)));
```
Comment (Aurel300) [inline prusti-encoder/src/lib.rs]: Can you explain this? Is this only to make the output contain these extra functions "within" the domain, even though they are actually allocated as part of another encoder? Could they be placed into a separate domain?

File: prusti-encoder/src/encoders/util.rs [hunk 15]
```diff
@@ -0,0 +1,31 @@
+use vir::VirCtxt;
+use prusti_rustc_interface::{
+    middle::{mir, ty},
+    span::def_id::DefId,
+};
+
+/// Returns input and output types of a function
+pub fn get_func_sig<'tcx>(vcx: &VirCtxt<'tcx>, def_id: DefId) -> (Vec<ty::Ty<'tcx>>, ty::Ty<'tcx>) {
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/util.rs]: Why this function? Why do something different with local bodies? And why is `fn_sig` generally not sufficient? I am quite hesitant to have something like this in the codebase, because (apart from another `util` file) it suggests that the type of a method is fully determined by the `DefId`. However, this is only the case in this generic encoding, but is not true for example when monomorphisation is used.

File: prusti-encoder/src/lib.rs (hunk 14 above)
Comment (zgrannan) [reply prusti-encoder/src/lib.rs]: > Is this only to make the output contain these extra functions "within" the domain, even though they are actually allocated as part of another encoder?

Yes this was exactly my intention

> Could they be placed into a separate domain?

Ah I didn't think about doing that. It seems like it's not necessary to include axioms / functions within the same domain (apart from readability, perhaps). Should I just delete this and make the encoder output its own `Domain`?

File: prusti-encoder/src/encoders/util.rs (hunk 15 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/util.rs]: I copied the logic from here: https://github.com/Aurel300/prusti-dev/blob/7aaa380c654bd17dada922f189a522f422c0604d/prusti-encoder/src/encoders/local_def.rs#L75-L105

I suppose it might be worth adding a comment there to explain what it's doing? I am not sure why there needs to be a case split there.

Agree with your points, I'll:
- remove this file
- change calls of this function to instead use `fn_sig`

I don't think it should cause any issues (I can't think of any reason why `fn_sig` wouldn't work)

File: prusti-encoder/src/lib.rs (hunk 14 above)
Comment (Aurel300) [reply prusti-encoder/src/lib.rs]: Yes, please remove this and output a separate domain. As one of your comments said, this will need to happen for output through JNI anyway.

**Comment by Aurel300** [comment]
Besides the two comments, there are some more problems when testing:

- Viper complains about some of the axioms, e.g. `make_generic_s_Bool is not a domain function`: this is because axioms cannot define the behaviour of non-domain functions. The `make_generic/concrete_*` functions should be domain functions (in yet another domain, I guess). The precondition of `make_concrete` can be dropped. We can re-introduce another caller function with a precondition later, but it also isn't super important: if the postcondition is defined via an axiom like this one
  ```viper
  axiom {
    forall self: s_Bool :: {make_generic_s_Bool(self)} make_concrete_s_Bool(make_generic_s_Bool(self)) == self
  }
  ```
  then other cases will simply be undefined and probably cause a program to fail to verify. This precondition would be more of a typecheck for ourselves anyway.
- I don't understand why there is a `typeof` function defined over values. When is it necessary? In the code I find a usage related to getting type parameters out of a snapshot of a generic. Why is it not possible to extract the types from the type information? I would expect to only have `Type -> Type` functions in the code, never `Snap -> Type`.
- The generated Viper code now contains an encoding for generic pure functions that seems correct, i.e. there are generic arguments etc, but there are still multiple `_CALLER` functions generated, which now end up containing the same body.

**Comment by zgrannan** [comment]
> * I don't understand why there is a `typeof` function defined over values. When is it necessary? In the code I find a usage related to getting type parameters out of a snapshot of a generic. Why is it not possible to extract the types from the type information? I would expect to only have `Type -> Type` functions in the code, never `Snap -> Type`.

Consider the struct `struct Pair<T, U>(T, U)`. We need to axiomatize that the 1st accessor returns a value of type `T` (and the second of type `U`). The axiom for the first would be:

```
axiom ax_s_Pair_read_0_type {
  forall self: s_Pair :: {s_Pair_read_0(self)} (typ(s_Pair_read_0(self))) == (s_Pair_typaram_T(typeof_s_Pair(self)))
}
```

Where `typ` returns the type of an `S_Param`.

I think it's also necessary for the `make_generic` functions, e.g,:

```
function make_generic_s_Pair(
  self: s_Pair
): s_Param
  ensures (typ(result)) == typeof_s_Pair(self))
```

**Comment by zgrannan** [comment]
> ~Viper complains about some of the axioms, e.g. `make_generic_s_Bool is not a domain function`: this is because axioms cannot define the behaviour of non-domain functions. The `make_generic/concrete_*` functions should be domain functions (in yet another domain, I guess). The precondition of `make_concrete` can be dropped.~

This seems to be okay in current Silicon

> * The generated Viper code now contains an encoding for generic pure functions that seems correct, i.e. there are generic arguments etc, but there are still multiple `_CALLER` functions generated, which now end up containing the same body.

Will make this change

**Comment by zgrannan** [comment]
@Aurel300 updated to remove the extra generated functions when we're not monomorphising.

Also, the monomorphic encoding of functions can be enabled via the flag `mono_function_encoding` in the `prusti-encoder` crate. I suppose in principle having this changable as a normal option (passed in as command-line argument or environment variable) wouldn't actually be very difficult, but I don't know if we want to change any infrastructure for Prusti options in v2 .

File: prusti-encoder/src/encoders/mir_pure_function.rs [hunk 16]
```diff
@@ -1,47 +1,62 @@
...
+    }
 
-#[derive(Clone, Debug)]
-pub struct MirFunctionEncOutput<'vir> {
-    pub function: vir::Function<'vir>,
+    fn mk_function_ident<'vir, 'tcx>(
+        vcx: &'vir vir::VirCtxt<'tcx>,
+        def_id: &Self::TaskKey<'tcx>,
+    ) -> vir::ViperIdent<'vir> {
+        vir::vir_format_identifier!(vcx, "f_{}_CALLER", vcx.tcx().item_name(*def_id))
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/mir_pure_function.rs]: I guess there is no reason for this to say "CALLER" anymore? IIUC that bit was there to separate what is being encoded from what is calling this.

File: prusti-encoder/src/encoder_traits/pure_function_enc.rs [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+}
+impl<'vir> task_encoder::OutputRefAny for MirFunctionEncOutputRef<'vir> {}
+
+#[derive(Clone, Debug)]
+pub struct MirFunctionEncOutput<'vir> {
+    pub function: vir::Function<'vir>,
+}
+
+/// Task encoders can implement this trait to encode pure functions. We use this
+/// code to support sharing code betwenn monomorphic and polymorhic encoding of
```
Comment (Aurel300) [inline prusti-encoder/src/encoder_traits/pure_function_enc.rs]:
```suggestion
/// code to support sharing code between monomorphic and polymorphic encoding of
```

File: prusti-encoder/src/encoders/type/predicate.rs [hunk 18]
```diff
@@ -61,22 +61,37 @@ pub enum PredicateEncData<'vir> {
...
     //pub method_refold: &'vir str,
     pub specifics: PredicateEncData<'vir>,
+    pub generics: &'vir [vir::LocalDecl<'vir>],
 }
 impl<'vir> task_encoder::OutputRefAny for PredicateEncOutputRef<'vir> {}
 
 impl<'vir> PredicateEncOutputRef<'vir> {
+    pub fn ref_to_args<'tcx>(
+        &self,
+        vcx: &'vir vir::VirCtxt<'tcx>,
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/type/predicate.rs:85]: Generally, I think we should avoid passing `vcx`: it can be acquired inside the function with `with_vcx`.

File: vir/src/debug_info.rs [hunk 19]
```diff
@@ -1,24 +1,25 @@
...
+        use std::marker::PhantomData;
     }
 }
 
 // Definition of `DebugInfo`
 cfg_if! {
     if #[cfg(feature="vir_debug")] {
         #[derive(Clone, Copy, Debug)]
-        pub struct DebugInfo(Option<NonNull<DebugInfoData>>);
+        pub struct DebugInfo<'a>(Option<&'a Mutex<DebugInfoData>>);
```
Comment (Aurel300) [inline vir/src/debug_info.rs:20]: `Mutex` is quite heavy-weight and hopefully unnecessary: we are not even multi-threaded. Why the change? Is this to avoid the unsafe initialisation? Would a cell variant (`RefCell`?) work?

**Review by Aurel300 — APPROVED** [review]
Apart from the small comments, LGTM.

File: prusti-encoder/src/encoders/type/predicate.rs (hunk 18 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/type/predicate.rs:85]: This function is called within a call to `with_vcx` sometimes, or where there's already a vcx in context, e.g

```rust
let inner_ref_to_args = inner.ref_to_args(
    self.vcx,
    lifted_ty,
    self_ref
);
```

Is there not much overhead / problems with nesting `with_vcx` calls? If so, I can remove this (and some other places that take an explicit vcx parameter).

File: prusti-encoder/src/encoders/mir_pure_function.rs (hunk 16 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/mir_pure_function.rs]: Removed it!

File: vir/src/debug_info.rs (hunk 19 above)
Comment (zgrannan) [reply vir/src/debug_info.rs:20]: The prusti-server integration requires that `DebugInfo` be sharable between threads (due to its inclusion in `vir::Expr` for example), hence the Mutex.

We could implement an unsafe `Send` trait for it, but using `NonNull` it would not actually be threadsafe. I don't think this would actually break though.

Another alternative I considered is to have a way to strip the debuginfo from the Expr, by making the type of DebugInfo parametric and writing a function converting `Expr<DebugInfo>` to `Expr<()>`. The problem is that this would further complicate the type signature (`ExprGen` already takes 3 parameters).

Comment (zgrannan) [reply vir/src/debug_info.rs:20]: The actual error when `DebugInfo` is not `Send` looks like this, btw:

```
175 | pub fn json<T: DeserializeOwned + Send>() -> impl Filter<Extract = (T,), Error = Rejection> + Copy {
    |                                   ^^^^ required by this bound in `json`

error[E0599]: the method `map` exists for struct `Map<Map<And<And<And<And<..., ...>, ...>, ...>, ...>, ...>, ...>`, but its trait bounds were not satisfied
  --> prusti-server/src/server.rs:80:10
   |
73 |       let json_verify = warp::path!("json" / "verify")
   |  _______________________-
74 | |         .and(warp::body::json())
75 | |         .map(init_vcx)
76 | |         .map(build_verification_request_handler(
...  |
79 | |         ))
80 | |         .map(|response| warp::reply::json(&response));
   | |         -^^^ method cannot be called due to unsatisfied trait bounds
   | |_________|
   ```
